controlplane/kubeadm/internal/proxy/dial.go

```diff
@@ -93,33 +94,50 @@ func (d *Dialer) DialContext(_ context.Context, network string, addr string) (ne
 
 	dialer := spdy.NewDialer(d.upgrader, &http.Client{Transport: d.proxyTransport}, "POST", req.URL())
 
-	p, _, err := dialer.Dial(portforward.PortForwardProtocolV1Name)
+	// Create a new connection from the dialer.
+	//
+	// Warning: Any early return should close this connection, otherwise we're going to leak them.
+	connection, _, err := dialer.Dial(portforward.PortForwardProtocolV1Name)
 	if err != nil {
 		return nil, errors.Wrap(err, "error upgrading connection")
 	}
+
+	// Create the headers.
 	headers := http.Header{}
-	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	// Set the header port number to match the proxy one.
 	headers.Set(corev1.PortHeader, fmt.Sprintf("%d", d.proxy.Port))
+
 	// We only create a single stream over the connection
 	headers.Set(corev1.PortForwardRequestIDHeader, "0")
-	errorStream, err := p.CreateStream(headers)
+
+	// We're not writing to the error stream, so close it.
+	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	errorStream, err := connection.CreateStream(headers)
 	if err != nil {
-		return nil, err
+		return nil, kerrors.NewAggregate([]error{
+			err,
+			connection.Close(),
+		})
 	}
-
 	if err := errorStream.Close(); err != nil {
-		return nil, err
+		return nil, kerrors.NewAggregate([]error{
+			err,
+			connection.Close(),
+		})
 	}
 
+	// Set the stream type header again to data stream, and create it.
```

Comment:
```suggestion
	// Create the data stream
```

Reply:
I'd keep what's there now, it was confusing at first reading the code we're using the same exact headers and then overwriting one, added the above for posterity

Reply:
Reworded a little